doc/developer/design/20230903_avro_doc.md

```diff
@@ -0,0 +1,322 @@
+list, where it can be freely intermixed with other options, e.g.:
+
+```sql
+FORMAT AVRO (
+    DOC ON ...,
+    CONFLUENT SCHEMA REGISTRY = '...',
+    DOC ON ...,
+)
+```
+
+Each `DOC ON` option describes a documentation comment that should be attached
+to a single field in the Avro schema. See [Avro schema node
+specifiers](#avro-schema-node-specifiers) below for details on how users name
+the Avro schema node on which to attach documentation.
+
+Materialize will reject `CREATE SINK` commands that specify multiple `DOC ON`
+options naming the same field.
+
+Materialize will automatically use [comments] on the columns of materialized
+views, or its constituent types, as the default documentation for a field.
+Specifically, for a given node of the Avro schema, Materialize will consider the
+following sources of documentation in order:
+
+  * A `DOC ON` option specifically naming that schema node.
+  * A comment on the column that produced the schema node. If the schema
+    element is at the top level, this will be a comment on a column of the
+    sink's materialized view. If the schema node is nested, this will be a
+    comment on a column of a record type.
+  * A comment on the type of the column that produced the schema node.
+
+For example, consider the following sink:
+
+```sql
+CREATE TYPE point AS (x integer, y integer);
+CREATE MATERIALIZED VIEW v AS SELECT ROW(1, 1)::point AS c1, 'text' AS c2;
+CREATE SINK FROM v INTO KAFKA ... FORMAT AVRO ...;
+```
+
+For `c1`, Materialize will consider the following sources of documentation
+in order:
+
+  * A `DOC ON` option naming `c1`.
+  * A comment on column `v.c1`.
+  * A comment on type `point`.
+
+`c2` is analogous.
+
+For `c1.x`, Materialize will consider the following sources of documentation
+in order:
+
+  * A `DOC ON` comment naming `c1.x`.
+  * A comment on column `point.x`.
+  * A comment on type `integer`.
+
+`c1.y` is analogous.
+
+Changes made to comments *after* the the sink is created will not be reflected
+in the Avro schema.
+
+#### Avro schema node specifiers
+
+[Brennan](https://github.com/umanwizard) invented a way of unambiguously
+naming Avro schema nodes in the `kgen` load generator. Its syntax and semantics
+are described here.
+
+An Avro schema node can appear in the following contexts:
+
+  1. The top-level node of the schema,
+  2. One of the fields of a record,
+  3. One of the branches of a union,
+  4. The value type of an array, or
+  5. The value type of a map.
+
+We define the "unambiguous name" of a schema node recursively as follows:
+
+ * If the node is of a [named
+   type](https://avro.apache.org/docs/1.10.2/spec.html#names) (record, enum, or
+   fixed) then its unambiguous name is the same as its "fullname" as defined by
+   the Avro spec; e.g. just its name if it is not part of a namespace, or its
+   namespace, followed by `.`, followed by its name if it is.
+ * Otherwise, if the node is the top-level node of the schema, its unambiguous
```

Comment:
Since we'll always have a top level record, of name either `envelope` (default) or user provided, then something like `DOC ON envelope` can be used as a top level comment? 

cc @sjwiesman

Reply:
Exactly! I just added that to the doc and pushed it up.